### DISCUSSION TITLE
Android ANE : 'bcprov-jdk18on-1.79.jar' using Jetifier. Reason: IllegalArgumentException, message: Unsupported class file major version 65.

### DISCUSSION THREAD

**Discussion by pol2095** [airsdk/Adobe-Runtime-Support#3620] [Q&A]
Hello,

I tried to compile an ANE using this jar dependency [bcprov-jdk18on-1.79.jar](https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk18on/1.79) but when when I try to compile my android app, I have this message error :
```
unexpected failure: Unable to run java: com.adobe.air.ADTException: gradle tool failed:
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeDebugJavaResource'.
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Failed to transform bcprov-jdk18on-1.79.jar to match attributes {artifactType=android-java-res, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for JetifyTransform: C:\Users\user\AppData\Local\Temp\4b74b8c5-1fbf-4cc7-a92f-c26d9ebad863\app\libs\bcprov-jdk18on-1.79.jar.
         > Failed to transform 'C:\Users\user\AppData\Local\Temp\4b74b8c5-1fbf-4cc7-a92f-c26d9ebad863\app\libs\bcprov-jdk18on-1.79.jar' using Jetifier. Reason: IllegalArgumentException, message: Unsupported class file major version 65. (Run with --stacktrace for more details.)
           Suggestions:
            - Check out existing issues at https://issuetracker.google.com/issues?q=componentid:460323&s=modified_time:desc, it's possible that this issue has already been filed there.
            - If this issue has not been filed, please report it at https://issuetracker.google.com/issues/new?component=460323 (run with --stacktrace and provide a stack trace if possible).

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 18s

java.io.IOException: Unable to run java: com.adobe.air.ADTException: gradle tool failed:
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeDebugJavaResource'.
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Failed to transform bcprov-jdk18on-1.79.jar to match attributes {artifactType=android-java-res, org.gradle.libraryelements=jar, org.gradle.usage=java-runtime}.
      > Execution failed for JetifyTransform: C:\Users\user\AppData\Local\Temp\4b74b8c5-1fbf-4cc7-a92f-c26d9ebad863\app\libs\bcprov-jdk18on-1.79.jar.
         > Failed to transform 'C:\Users\user\AppData\Local\Temp\4b74b8c5-1fbf-4cc7-a92f-c26d9ebad863\app\libs\bcprov-jdk18on-1.79.jar' using Jetifier. Reason: IllegalArgumentException, message: Unsupported class file major version 65. (Run with --stacktrace for more details.)
           Suggestions:
            - Check out existing issues at https://issuetracker.google.com/issues?q=componentid:460323&s=modified_time:desc, it's possible that this issue has already been filed there.
            - If this issue has not been filed, please report it at https://issuetracker.google.com/issues/new?component=460323 (run with --stacktrace and provide a stack trace if possible).

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 18s

        at com.adobe.air.apk.AABOutputStream.buildGradle(AABOutputStream.java)
        at com.adobe.air.apk.AABPackager.renameOutput(AABPackager.java)
        at com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
        at com.adobe.air.apk.AABPackager.createPackage(AABPackager.java)
        at com.adobe.air.ADT.parseArgsAndGo(ADT.java)
        at com.adobe.air.ADT.run(ADT.java)
        at com.adobe.air.ADT.main(ADT.java)
```
Thanks

**Comment by ajwfrost** [comment]
It doesn't look like this is related to the BouncyCastle library, it's more about the Android build tools I think...

We have seen this a few times now; can you see whether the below suggestion helps? i.e. forcing it to use a later Gradle version?
https://github.com/airsdk/Adobe-Runtime-Support/issues/2965#issuecomment-1857436071

thanks

**Reply by ajwfrost** [reply]
Hi @pol2095  - I saw your response on #2965, can I check what versions of the Android build tools and SDK you have on your computer, and whether you have any "targetSdk" value in your application descriptor manifestAdditions? I think this is all related to the different versions of the tools and libraries that it uses... something, somewhere, seems to want to use class file version 65 = Java 21.

It may also be possible to debug this sort of issue by creating this with the "KeepAndroidStudioOutput" configuration flag turned on, and then open up the generate project in Android Studio (or just change your target platform to "android-studio" rather than "apk"). If you can build from Android Studio first time then we're doing something wrong I think! But it should also be possible to run the gradle build command that we use in ADT, and see what then happens (and it can then be run with call stack and debug information output..)

Ideally though, if you can confirm the various settings, we can see if we can reproduce it here; failing that, we can perhaps look at doing some kind of ADT update with some jar-scanning code added to try to find out if there really is a Java class file within the application files that's got this Java requirement...

thanks

**Reply by pol2095** [reply]
I used the same agp version and gradle version that use to create the jar on Android Studio, it works now :
```
<gradleVersion>8.9</gradleVersion>
<androidGradlePluginVersion>8.7.3</androidGradlePluginVersion>
```
---------------

I founded a workaround to work with Gradle 8.0 using `bcprov-jdk18on-1.79.jar`, I deleted `META-INF\versions\21` in the jar and it work correctly

![21](https://github.com/user-attachments/assets/0b67bc81-c2df-4c34-b20c-236173694cdb)

**Reply by ajwfrost** [reply]
Oh interesting! So that folder is described in the [spec](https://docs.oracle.com/en/java/javase/21/docs/specs/jar/jar.html#the-meta-inf-directory) as:
> This directory contains underneath it versioned class and resource files for a [multi-release](https://docs.oracle.com/en/java/javase/21/docs/specs/jar/jar.html#multi-release-jar-files) JAR file.

Which basically implies that the Android build tools can't cope with this spec, at least until you hit the Gradle/AGP version you mention above.

Thanks for working this one out, good to know!
